### PR TITLE
invalidate watches when client reconnects

### PR DESCRIPTION
**Pull request by sjpotter**
also provide API for other wrapper clients to also create invalid watch states programatically.

Reasoning: if a user does a WATCH and then the client reconnects, the watch is no longer active, but if a user does a MULTI/EXEC after that, they wont know, and since the WATCH is no longer active, the request has no protection.

The API is needed for when a wrapper client (say sentinel, cluster) might close the underlying client and reopen a new one transparently to the user.  Just like in the reconnection case, this should result in an error, but its up to the wrapping client to provide the appropriate error

### Description

> Describe your pull request here

---

### Checklist

- [ ] Does `npm test` pass with this change (including linting)?
- [ ] Is the new or changed code fully tested?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?

### CODE REVIEW

**Comment by sjpotter** [comment]
this PR is how I imagine the API as needed, need to remove WATCH/UNWATCH command files probably now.

File: packages/client/lib/client/index.ts [hunk 1]
```diff
@@ -279,6 +279,9 @@ export default class RedisClient<
   #monitorCallback?: MonitorCallback<TYPE_MAPPING>;
   private _self = this;
   private _commandOptions?: CommandOptions<TYPE_MAPPING>;
+  #dirtyWatch?: string;
+  #epoch: number;
```
Comment (sjpotter) [inline packages/client/lib/client/index.ts:283]: does it default to Zero? don't see it set besides the ++ on 'ready'

Comment (sjpotter) [reply packages/client/lib/client/index.ts:284]: wondering if the epoch should be moved into the RedisSocket and not in the client itself.

Comment (leibale) [reply packages/client/lib/client/index.ts:284]: 100% agree with you

Comment (leibale) [reply packages/client/lib/client/index.ts]: fixed

**Review by leibale — APPROVED** [review]

**Comment by sjpotter** [comment]
LGTM

**Pull request closed without merging**